docs/debug-adapter.md

```diff
@@ -0,0 +1,37 @@
+---
+id: dap-integration
+title: Debug Adapter Protocol Integration 
+sidebar_label: DAP Integration guide
+---
+
+## Debug Adapter Protocol integration guide
+
+Bloop allows the client to run and debug both the main classes and the tests defined in the project. 
+This is possible thanks to the Debug Adapter Protocol, which bloop exposes to the client 
+in a run/test task start notification. This adapter expects either a 
+`ScalaTestClassesItem` or `ScalaMainClass` as a configuration parameter for the launch request.
+
+### Client-Server communication
+![communication sequence diagram](assets/dap-sequence.png)
+
+1. BSP Client chooses which main class to run. To do so, it requests the list of all runnable main classes from the BSP Server.
+2. BSP Client request the chosen class to be run.
+3. BSP Server sets up a new Debug Adapter. It has all the details for running, since those came in the request data field. If the request from 2. had a field debug set to true, the Adapter will be launched with full debugging capabilities. Otherwise only lifecycle-related endpoints (launch, restart, terminate etc.) will be supported.
+4. BSP Server notifies the BSP Client of the Debug Adapter uri
+5. BSP Client notifies the Debug Client to connect to the given uri
+6. Debug Client converses with the Debug Adapter
+7. Debug Adapter notifies the BSP Server of its progress. As a result it sends notifications to the Client
+8. Debug Adapter notifies the BSP Server it is about to terminate
+9. BSP Server notifies the BSP Client the run task has finished
+10. BSP Server responds to the BSP Client with the run result of request from 2.
+
+### Example: [metals](https://github.com/scalameta/metals) integration
+
+Metals is an interesting case of BSP client. 
+It is acting as a language server therefore it must allow the language client to handle the Debug Client. 
+
+![metals communication sequence diagram](assets/dap-example-metals.png)
+1. Metals caches the available main classes.
+2. User opens a source file. As a result, Language Client asks metals for code lenses, which are resolved using the cached main classes metadata.
+3. User clicks on the code lens with the intention to run the specific main class. Metals requests a Debug Adapter uri from the BSP Server and then passes it back to the Language Client.
+4. Language Client instruments the Debug Client to connect to this Debug Adapter
```

Comment:
@olafurpg  asked:
> What will be the arguments of the `launch` request via DAP? I think it would be great if we can build on top of the launch.json configuration in VS Code so that users can trigger run/test via keyboard shortcuts instead of clicking on "run/test" buttons.
> 
> The nice thing about launch.json configuration is that it could committed into git and shared across team members.

Reply:
Let me rephrase the problem and suggest a solution. Then we can discuss whether it solves your concerns and if it is viable at all.

Problem: 
1. Configurations must be visible from the outside of BSP/metals (in Language Client)
2. Configurations created/updated bythe Lanugage Client must remain usable

Propositions:
### configuration storage in metals

This storage could be an in-memory cache (populated on startup from client's `launch.json`) or a directory maintained by metals (but that is harder, I think). 

Whenever Language Client configuration store is updated (onDidSave for `launch.json`?), metals parses this file and updates its internal storage.

Whenever a configuration is run and is not present in this storage, metals appends the configuration to the `launch.json`. This configuration is a `scalaMainClass` or `scalaTestClassesItem` and is saved without any changes (meaning: as it was sent by the BSP Server earlier).

How will we know it is being run for the first time? We can compare checksums or something. That way we will persist every configuration that has been launched, which will allow the user to repeatedly run same configurations without using code lenses. Also, main classes never launched will not be stored. Another bonus is that now we can store test configuration which spans across many build targets (let's say in metals `slow` and `unit` but not `cross`).
	
The whole process would work like that (sorry for the lack of diagrams ;P):
	1. metals fetches main classes
	2. vs code requests code lenses
	3. user clicks on the code lens
	4. vs code sends a `metals-main-run` request with the configuration
	5. metals checks-up its internal cache of configurations (magic hashing techniques). If the configuration is not persisted in Language Client configuration cache, metals appends it. Then, regardless of whether it was saved or not, metals requests `buildTarget/run`
	6. user makes some changes to this configuration and runs it.
	7. vs code resolves the configuration by name and sends it to metals via `metals-main-run`. Metals checks if it has such configuration stored. It doesn't, since the user created it himself by modifying. So, metals just runs `buildTarget/run`.
	8. user clicks code lens once more. metals does not recognise this configuration and therefore appends it. Now there are two configurations in `launch.json`: modified by user and "bare" from clicking the code lens.

### configuration store in bloop

I have not explored this approach a lot but also I am not sure storing configurations fits into the range of responsibilities for the BSP Server. Thoughts?

Reply:
First, let me try to define a few objectives that I think are desirable for any solution around managing launch configurations:

- the user should be able to trigger run/test via `launch.json` using keyboard shortcuts, it shouldn't be necessary to use a mouse to click on a code lens
- the user should be able to commit `launch.json` to version control so that other team members can checkout a clean repo to reuse custom launch configurations

Here's a proposal for how we could manage launch configurations:

- no configuration storage in Bloop, Bloop only defines the spec for what launch configurations it supports (build target identifier, main class, test class, link options, etc.) and executes the launch configurations it receives via DAP `launch`.
- Metals stores all executed launch configuration in the `.metals/metals.db` SQL database

```sql
create table launch_configuration(
  when_launched timestamp,
  launch_data: varchar not null, -- JSON encoded string.
  -- Up for discussion, human readable description of the launch configuration.
  -- For example "Test UserSuite in core-test".
  display_name varchar not null
);
```
- Metals auto-generates a single launch configuration in VS Code `launch.json` with value `{"name": "Run Last Task"}`, which executes the persisted launch configuration with the most recent timestamp. No hashing or advanced logic is needed to generate this config because the JSON is so simple.
- Metals registers the LSP server command "Metals: Save Recent Task", which opens a select list showing the last 5-10 executed launch configurations. The configuration selected by the user will be persisted in the VS Code `launch.json` file, with full details including build target identifier and test/main class.
- Metals registers the LSP server command "Metals: Run Recent Task", which works the same as "Save Recent Task" except it starts a debug session to launch that configuration instead of saving it to launch.json

Reply:
> - the user should be able to trigger run/test via launch.json using keyboard shortcuts, it shouldn't be necessary to use a mouse to click on a code lens

That is the responsibility of the editor - to make configurations available to the user under the specified key binding. Especially since the editor has the full knowledge of the configurations. Currently, I could not make it work with the built-in `run last task` command - @jvican you mentioned something about it working, if I remember correctly?

Anyway, if it is in fact impossible to run a specific configuration in a convenient way (from the config file, using commands), then it probably merits an issue in the vs code. 

> - the user should be able to commit `launch.json` to version control so that other team members can checkout a clean repo and still reuse custom launch configurations

Agreed.

All in all, seems doable. But it is for sure more convenient for the developer - but is it more convenient for the user?

Reply:
> is it more convenient for the user?

I think so, I tried to design for the user experience first and architecture (implementation details) second.

> the user should be able to trigger run/test via launch.json using keyboard shortcuts, it shouldn't be necessary to use a mouse to click on a code lens

Code lenses

Reply:
> the user should be able to trigger run/test via launch.json using keyboard shortcuts, it shouldn't be necessary to use a mouse to click on a code lens

I agree the keyboard shortcut part is a responsibility of the editor. Let me try to reformulate the point: it should be possible to launch configurations through `launch.json` configurations in addition to code lenses because launch configurations are the standard interface to run Python/Node.js/Java/C# applications in VS Code.

Reply:
Well it surely depends on ones habits, so let's wait for the opinion of others

Reply:
The primary interface that VS Code uses for launching applications is `launch.json`, which can be selected like this
![2019-05-27 12 18 15](https://user-images.githubusercontent.com/1408093/58413611-8a3a8580-8079-11e9-8031-b28342513c74.gif)

Me and @jvican looked at how it works in the Python client and they use the same interface for debugging. For regular running without breakpoint debugging they use a separate custom interface, which is confusing in my opinion.

<img width="1161" alt="Screenshot 2019-05-24 at 21 34 46" src="https://user-images.githubusercontent.com/1408093/58413701-bd7d1480-8079-11e9-9377-486a4c3e7983.png">